### PR TITLE
Cleanup network usage in NBitcoin #1

### PR DESCRIPTION
**Pull request by fassadlr**
We need to start removing Network.Main references from the NBitcoin project as we eventually want to have all the networks int heir own project.

### CODE REVIEW

File: src/NBitcoin/ConsensusFactory.cs [hunk 1]
```diff
@@ -39,6 +39,13 @@ public class ConsensusFactory
         /// </summary>
         private readonly TypeInfo transactionType = typeof(Transaction).GetTypeInfo();
 
+        private readonly Network network;
```
Comment (dangershony) [inline src/NBitcoin/ConsensusFactory.cs]: In this case, please remove this field.
`public Consensus Consensus { get; set; }`

File: src/NBitcoin/BlockStake.cs [hunk 2]
```diff
@@ -34,9 +34,9 @@ public BlockStake()
         {
         }
 
-        public BlockStake(byte[] bytes)
+        public BlockStake(Network network, byte[] bytes)
```
Comment (dangershony) [inline src/NBitcoin/BlockStake.cs]: We should probably not do that, ther eshould be a Load() method instead.

File: src/NBitcoin/IBitcoinSerializable.cs [hunk 3]
```diff
@@ -10,23 +10,23 @@ public interface IBitcoinSerializable
 
     public static class BitcoinSerializableExtensions
     {
-        public static void ReadWrite(this IBitcoinSerializable serializable, Stream stream, bool serializing, ProtocolVersion version = ProtocolVersion.PROTOCOL_VERSION, Network network = null)
+        public static void ReadWrite(this IBitcoinSerializable serializable, Stream stream, bool serializing, ConsensusFactory consensusFactory, ProtocolVersion version = ProtocolVersion.PROTOCOL_VERSION)
```
Comment (dangershony) [inline src/NBitcoin/IBitcoinSerializable.cs]: I think maybe perhaps we should create a  default `ConsensusFactory` that will just return null (or throw when one of the types it expect is asked) for every type.
then this method can be called with no `ConsensusFactory` and still work.

File: src/NBitcoin/ConsensusFactory.cs (hunk 1 above)
Comment (fassadlr) [reply src/NBitcoin/ConsensusFactory.cs]: `Consensus` still gets used in the `Coins` class to check the `IsProofOfStake` property. Maybe we can look at removing this or find a work around in the next PR?

File: src/NBitcoin/BlockStake.cs (hunk 2 above)
Comment (fassadlr) [reply src/NBitcoin/BlockStake.cs]: OK yeah I have created a `BlockStake.Load` that takes in either `byte[]` and `Block`.

File: src/NBitcoin/IBitcoinSerializable.cs (hunk 3 above)
Comment (fassadlr) [reply src/NBitcoin/IBitcoinSerializable.cs]: Hmmm I'm really not a fan of these methods that can be called with null defaults. I think in the solution this is something that we should be moving away from. I would much rather create another `ReadWrite` where we don't pass in a `ConsensusFactory`, should the need arise. This way we are more explicit as to the intent of the method.

File: src/NBitcoin/BlockStake.cs (hunk 2 above)
Comment (bokobza) [reply src/NBitcoin/BlockStake.cs]: why Load?

Comment (fassadlr) [reply src/NBitcoin/BlockStake.cs]: @bokobza This is to keep the pattern consistent throughout. We load blocks and block headers like this. Also, it lessens the footprint that the `Network` class has in NBitcoin. We can rename it to something else?

File: src/NBitcoin/IBitcoinSerializable.cs (hunk 3 above)
Comment (dangershony) [reply src/NBitcoin/IBitcoinSerializable.cs]: But there are many tyoes that use the serializer apart for the (Header,Block,Transaction) those dont need the factory.

File: src/NBitcoin/BlockStake.cs [hunk 4]
```diff
@@ -166,6 +180,11 @@ public PosTransaction(byte[] bytes) : this()
     /// </summary>
     public class PosConsensusFactory : ConsensusFactory
     {
+        public PosConsensusFactory(Network network)
```
[the remark this answers is not in the document]
Comment (fassadlr) [reply src/NBitcoin/BlockStake.cs]: Will remove 👍

File: src/NBitcoin/IBitcoinSerializable.cs (hunk 3 above)
Comment (fassadlr) [reply src/NBitcoin/IBitcoinSerializable.cs]: Yes I understand that :) but I think that the part you are referring to is here:

https://github.com/stratisproject/StratisBitcoinFullNode/blob/master/src/NBitcoin/BitcoinStream.cs#L194

So here, if we are serializing/deserializing an object that is NOT Header,Block,Transaction we will just use the Activator?

Why would all the tests pass and this branch be able to sync if this was an issue? (Just asking)

:)

**Comment by fassadlr** [comment]
@dangershony staking still works as per normal as well as all the integration tests pass 👍

File: src/NBitcoin/BIP32/BitcoinExtKey.cs [hunk 5]
```diff
@@ -67,7 +67,7 @@ public ExtKey ExtKey
                 if(this._Key == null)
                 {
                     this._Key = new ExtKey();
-                    this._Key.ReadWrite(this.vchData);
+                    this._Key.ReadWrite(this.vchData, network: this.Network);
```
Comment (dangershony) [inline src/NBitcoin/BIP32/BitcoinExtKey.cs]: After the change we made with DefaultConsensusFactory we don't need network in this code

File: src/NBitcoin/Consensus.cs [hunk 6]
```diff
@@ -101,7 +101,7 @@ public BIP9DeploymentsArray()
             }
         }
 
-        public Consensus()
+        public Consensus(Network network)
```
Comment (dangershony) [inline src/NBitcoin/Consensus.cs]: I think you can remove Network form the constructor

File: src/NBitcoin/IBitcoinSerializable.cs [hunk 7]
```diff
@@ -55,9 +68,9 @@ public static int GetSerializedSize(this IBitcoinSerializable serializable, Prot
             return GetSerializedSize(serializable, version, SerializationType.Disk);
         }
 
-        public static void ReadWrite(this IBitcoinSerializable serializable, byte[] bytes, ProtocolVersion version = ProtocolVersion.PROTOCOL_VERSION, Network network = null)
+        public static void ReadWrite(this IBitcoinSerializable serializable, byte[] bytes, Network network, ProtocolVersion version = ProtocolVersion.PROTOCOL_VERSION)
```
Comment (dangershony) [inline src/NBitcoin/IBitcoinSerializable.cs]: You probably want to create another method that does not need network (and instead use DefaultConsensuFactory).

File: src/NBitcoin/Script.cs [hunk 8]
```diff
@@ -409,14 +409,14 @@ public Script(IEnumerable<byte> data)
             this._Script = data.ToArray();
         }
 
-        public Script(byte[] data, bool compressed)
+        public Script(Network network, byte[] data, bool compressed)
```
Comment (dangershony) [inline src/NBitcoin/Script.cs]: Script will never get any of the types that are in ConsensusFactory so we dont need to pass network in this case)

Use the override without network when calling `compressor.ReadWrite`

File: src/NBitcoin/DefaultConsensusfactory.cs [hunk 9]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+
+namespace NBitcoin
+{
+    /// <summary>
+    /// A default object factory to create instances that is not block, block header or transaction.
+    /// </summary>
+    public sealed class DefaultConsensusFactory : ConsensusFactory
```
Comment (dangershony) [inline src/NBitcoin/DefaultConsensusfactory.cs]: Super duper this makes our serializer much easier to use

File: src/NBitcoin/BIP32/ExtPubKey.cs [hunk 10]
```diff
@@ -72,11 +72,11 @@ internal ExtPubKey()
         {
         }
 
-        public ExtPubKey(byte[] bytes)
+        public ExtPubKey(Network network, byte[] bytes)
```
Comment (dangershony) [inline src/NBitcoin/BIP32/ExtPubKey.cs]: I think we can remove this from the constructor `ExtPubKey` is not one of the types in consensus factory.

**Pull request merged**